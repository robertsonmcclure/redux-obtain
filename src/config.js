import _ from "lodash"

const requestHeaderSelector = _.memoize(token => ({
    "Content-Type": "application/json",
    Authorization: `Basic ${token}`
}))
export let config = {
    requestHeaderSelector: state => requestHeaderSelector(state.authentication.token),
    reduxStoreName: "resources",
    paginationInitialLoadLimit: 500,
    getOrderBys: props =>
        props && props.sortBy && props.sortDirection
            ? props.sortBy.map((column, index) => ({
                  column,
                  direction: props.sortDirection[index]
              }))
            : []
}

export const setupFetcher = options => {
    config = {
        ...config,
        ...options
    }
}
