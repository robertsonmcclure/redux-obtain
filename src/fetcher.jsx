import React, { Component } from "react"
import { connect } from "react-redux"
import * as actions from "./actions"
import { config } from "./config"
import Promise from "bluebird"
import _ from "lodash"
import getDisplayName from "react-display-name"
import qs from "querystring"

Promise.config({ cancellation: true })

/* valid props: {
        name,
        endpoint,
        method,
        acceptResponse,
        persistResource,
        requestBodySelector,
        defaultOrderBys,
        paginationKey
    }
 */

export const fetcher = (props, extraActions) => WrappedComponent =>
    connect(
        (state, ownProps) => {
            const mergedProps = { ...ownProps, ...props }
            const { name, endpoint, requestBodySelector } = mergedProps

            return {
                ...mergedProps,
                endpoint: typeof endpoint === "function" ? endpoint(state, ownProps) : endpoint,
                requestHeader: config.requestHeaderSelector(state),
                resource: state[config.reduxStoreName][name],
                requestBody: requestBodySelector && requestBodySelector(state, ownProps)
            }
        },
        {
            ...extraActions,
            ...actions
        }
    )(
        class Fetcher extends Component {
            static displayName = `Fetcher(${getDisplayName(WrappedComponent)})`

            componentDidMount() {
                this.props.addResource(this.props.name, this.props.paginationKey)
                this.sendNetworkRequest({
                    limit: config.paginationInitialLoadLimit,
                    offset: 0,
                    orderBys: config.getOrderBys(this.props.defaultOrderBys),
                    firstLoad: true,
                    requestBody: this.props.requestBody
                })
            }
            componentWillUnmount() {
                this.networkRequest && this.networkRequest.cancel()
                !this.props.persistResource && this.props.removeResource(this.props.name)
            }
            componentDidUpdate(prevProps) {
                if (
                    !_.isEqual(prevProps.requestBody, this.props.requestBody) ||
                    !_.isEqual(prevProps.endpoint, this.props.endpoint)
                ) {
                    this.networkRequest && this.networkRequest.cancel()
                    this.sendNetworkRequest({
                        endpoint: this.props.endpoint,
                        limit: config.paginationInitialLoadLimit,
                        offset: 0,
                        orderBys: config.getOrderBys(this.props.defaultOrderBys),
                        firstLoad: true,
                        requestBody: this.props.requestBody
                    })
                }
            }

            sendNetworkRequest = ({ limit, offset, orderBys, firstLoad, endpoint }) => {
                firstLoad && this.props.requestResource(this.props.name)
                const sort = orderBys.map(
                    item => `${item.direction === "DESC" ? "-" : ""}${item.column}`
                )
                const paginationQuery = this.props.paginationKey
                    ? qs.stringify({ limit, offset, sort })
                    : ""
                const requestMethod = this.props.paginationKey ? "GET" : this.props.method || "GET"
                const requestEndpoint = endpoint || this.props.endpoint
                this.networkRequest = new Promise((res, rej) =>
                    fetch(
                        `${requestEndpoint}${
                            requestEndpoint.indexOf("?") === -1 ? "?" : "&"
                        }${paginationQuery}`,
                        {
                            method: requestMethod,
                            headers: this.props.requestHeader,
                            credentials: "include",
                            body:
                                requestMethod !== "GET"
                                    ? JSON.stringify({
                                          ...this.props.requestBody
                                      })
                                    : undefined
                        }
                    )
                        .then(x => res(x))
                        .catch(e => rej(e))
                )
                    .then(res => {
                        if (res.status === 200) {
                            return res
                                .json()
                                .then(data =>
                                    firstLoad
                                        ? this.props.fetchSuccess(
                                              this.props.name,
                                              data,
                                              this.props.acceptResponse
                                          )
                                        : this.props.fetchAdditionalSuccess(
                                              this.props.name,
                                              data,
                                              this.props.acceptResponse
                                          )
                                )
                        } else {
                            return res
                                .text()
                                .then(error => this.props.fetchError(this.props.name, error))
                        }
                    })
                    .catch(e => console.error(e))
                return this.networkRequest
            }
            loadMoreRows = (indices, ui) => {
                const { startIndex, stopIndex } = indices
                return this.sendNetworkRequest({
                    limit: stopIndex - startIndex,
                    offset: startIndex,
                    orderBys: config.getOrderBys(ui),
                    firstLoad: false
                })
            }
            loadInitialRows = props => {
                return this.sendNetworkRequest({
                    limit: config.paginationInitialLoadLimit,
                    offset: 0,
                    orderBys: config.getOrderBys(props),
                    firstLoad: true
                })
            }
            render() {
                const {
                    name,
                    endpoint,
                    method,
                    acceptResponse,
                    persistResource,
                    requestBodySelector,
                    defaultOrderBys,
                    paginationKey,
                    resource,
                    ...pt
                } = this.props
                const fetcher = _.pickBy({
                    name,
                    endpoint,
                    method,
                    acceptResponse,
                    persistResource,
                    requestBodySelector,
                    defaultOrderBys,
                    paginationKey
                })

                delete pt.addResource
                delete pt.requestResource
                delete pt.fetchSuccess
                delete pt.fetchAdditionalSuccess
                delete pt.fetchError
                delete pt.removeResource
                delete pt.requestBody
                delete pt.requestHeader
                const paginationFunctions = paginationKey
                    ? {
                          _loadMoreRows: this.loadMoreRows,
                          _loadInitialRows: this.loadInitialRows,
                          totalSize: resource && resource.data.totalCount
                      }
                    : undefined

                return resource ? (
                    <WrappedComponent
                        {...pt}
                        {...resource}
                        paginationFunctions={paginationFunctions}
                        fetcher={fetcher}
                    />
                ) : (
                    <div />
                )
            }
        }
    )
