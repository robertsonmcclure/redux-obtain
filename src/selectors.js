import { config } from "./config"

export const getResourceData = (name, selector) => state => {
    const resource = state[config.reduxStoreName][name]
    if (!resource) return undefined
    return selector ? selector(resource.data) : resource.data
}

export const getMetaData = name => state => {
    const { loading, error } = state[config.reduxStoreName][name] || {}
    return { loading, error }
}
