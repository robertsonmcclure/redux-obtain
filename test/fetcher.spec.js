import React from "react"
import { Provider } from "react-redux"
import { mount, shallow } from "enzyme"
import toJson from "enzyme-to-json"
import configureStore from "redux-mock-store"
import { fetcher } from "../src"

const mockStore = configureStore()
Date.now = jest.fn(() => 42)

const initialStore = {
    authentication: { token: "dummy_token" },
    resources: {
        NAME: {
            data: {},
            loading: false,
            error: false
        }
    },
    stuff: { key: "zero" }
}

const wait = ms => {
    return new Promise(resolve => {
        setTimeout(resolve, ms)
    })
}

describe("Fetcher", () => {
    beforeEach(() => {
        fetch.mockClear()
        fetch = jest.fn().mockImplementation(
            () =>
                new Promise((resolve, reject) => {
                    resolve({
                        ok: true,
                        status: 200,
                        json: () => new Promise((res, rej) => res({ pKey: [1, 2, 3, 4] })),
                        text: () => new Promise((res, rej) => res({ what: "what" }))
                    })
                })
        )
    })
    it("should render empty div if store gets messed with", () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: "/api/endpoint",
            method: "GET"
        })(() => <div />)
        const store = mockStore({ ...initialStore, resources: {} })
        const wrapper = mount(
            <Provider store={store}>
                <Container passthrough="attribute" />
            </Provider>
        )
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it("should mount, call fetch and dispatch actions", () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: "/api/endpoint",
            method: "GET"
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(
            <Provider store={store}>
                <Container passthrough="attribute" />
            </Provider>
        )
        expect(toJson(wrapper)).toMatchSnapshot()
        expect(store.getActions()).toMatchSnapshot()
        expect(fetch.mock.calls).toMatchSnapshot()
    })
    it("should dispatch unmount action on unmount", () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: "/api/endpoint",
            method: "GET"
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(
            <Provider store={store}>
                <Container />
            </Provider>
        )
        wrapper.unmount()
        expect(store.getActions()).toMatchSnapshot()
        expect(fetch.mock.calls).toMatchSnapshot()
    })
    it("should dispatch request on props change", () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "GET",
            requestBodySelector: (_, props) => ({ stuff: props.stuff })
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(<Container store={store} stuff={{ one: "two" }} />)
        wrapper.setProps({ stuff: { four: "five" } })
        expect(fetch.mock.calls).toMatchSnapshot()
        expect(store.getActions()).toMatchSnapshot()
        wrapper.setProps({ hi: "shouldnt do anything" })
        expect(fetch.mock.calls).toMatchSnapshot()
        expect(store.getActions()).toMatchSnapshot()
    })
    it("should dispatch success", async () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "POST",
            requestBodySelector: (_, props) => ({ stuff: props.stuff })
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(<Container store={store} stuff={{ one: "two" }} />)
        await wait(1000)
        expect(store.getActions()).toMatchSnapshot()
    })
    it("should dispatch error", async () => {
        fetch = jest.fn().mockImplementation(
            () =>
                new Promise((resolve, reject) => {
                    resolve({
                        ok: true,
                        status: 400,
                        json: () => new Promise((res, rej) => res({ test: "this" })),
                        text: () => new Promise((res, rej) => res({ server: "error" }))
                    })
                })
        )
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "GET",
            requestBodySelector: (_, props) => ({ stuff: props.stuff })
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(<Container store={store} stuff={{ one: "two" }} />)
        await wait(1000)
        expect(store.getActions()).toMatchSnapshot()
    })
    it("should console.error other error", async () => {
        console.error = jest.fn()
        fetch = jest.fn().mockImplementation(
            () =>
                new Promise((resolve, reject) => {
                    reject({
                        message: "Some Crazy Error"
                    })
                })
        )
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "GET",
            requestBodySelector: (_, props) => ({ stuff: props.stuff })
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(<Container store={store} stuff={{ one: "two" }} />)
        await wait(1000)
        expect(console.error.mock.calls).toMatchSnapshot()
    })
    it("should use pagination", async () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "POST",
            paginationKey: "pKey"
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = mount(<Container store={store} stuff={{ one: "two" }} />)
        await wait(1000)
        expect(fetch.mock.calls).toMatchSnapshot()
    })
    it("should use loadMoreRows", async () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "POST",
            paginationKey: "pKey"
        })(() => <div />)
        const store = mockStore(initialStore)
        const wrapper = shallow(<Container store={store} stuff={{ one: "two" }} />)
        await wait(1000)
        wrapper
            .dive()
            .dive()
            .instance()
            .loadMoreRows({ startIndex: 100, stopIndex: 250 })
        expect(fetch.mock.calls).toMatchSnapshot()
        await wait(1000)
        expect(store.getActions()).toMatchSnapshot()
    })
    it("should use loadInitialRows with different orderBys", async () => {
        const Container = fetcher({
            name: "NAME",
            endpoint: state => `/api/${state.stuff.key}`,
            method: "POST",
            paginationKey: "pKey"
        })(() => <div />)

        const store = mockStore(initialStore)
        const wrapper = shallow(<Container store={store} stuff={{ one: "two" }} />)
        await wait(1000)
        wrapper
            .dive()
            .dive()
            .instance()
            .loadInitialRows({ sortBy: ["columnKey"], sortDirection: ["ASC"] })
        expect(fetch.mock.calls).toMatchSnapshot()
    })
})
