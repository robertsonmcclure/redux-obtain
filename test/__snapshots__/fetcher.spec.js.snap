// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetcher should console.error other error 1`] = `
Array [
  Array [
    Object {
      "message": "Some Crazy Error",
    },
  ],
]
`;

exports[`Fetcher should dispatch error 1`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "error": Object {
      "server": "error",
    },
    "name": "NAME",
    "type": "@@fetcher/FETCH_ERROR",
  },
]
`;

exports[`Fetcher should dispatch request on props change 1`] = `
Array [
  Array [
    "/api/zero?",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
  Array [
    "/api/zero?",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Fetcher should dispatch request on props change 2`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
]
`;

exports[`Fetcher should dispatch request on props change 3`] = `
Array [
  Array [
    "/api/zero?",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
  Array [
    "/api/zero?",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Fetcher should dispatch request on props change 4`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
]
`;

exports[`Fetcher should dispatch success 1`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "acceptResponse": undefined,
    "data": Object {
      "pKey": Array [
        1,
        2,
        3,
        4,
      ],
    },
    "name": "NAME",
    "type": "@@fetcher/FETCH_SUCCESS",
  },
]
`;

exports[`Fetcher should dispatch unmount action on unmount 1`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REMOVE_RESOURCE",
  },
]
`;

exports[`Fetcher should dispatch unmount action on unmount 2`] = `
Array [
  Array [
    "/api/endpoint?",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Fetcher should mount, call fetch and dispatch actions 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Fetcher(Unknown))
    passthrough="attribute"
  >
    <Fetcher(Unknown)
      addResource={[Function]}
      endpoint="/api/endpoint"
      fetchAdditionalSuccess={[Function]}
      fetchError={[Function]}
      fetchSuccess={[Function]}
      method="GET"
      modifyResource={[Function]}
      name="NAME"
      passthrough="attribute"
      removeResource={[Function]}
      requestHeader={
        Object {
          "Authorization": "Basic dummy_token",
          "Content-Type": "application/json",
        }
      }
      requestResource={[Function]}
      resource={
        Object {
          "data": Object {},
          "error": false,
          "loading": false,
        }
      }
    >
      <Component
        data={Object {}}
        error={false}
        fetcher={
          Object {
            "endpoint": "/api/endpoint",
            "method": "GET",
            "name": "NAME",
          }
        }
        loading={false}
        modifyResource={[Function]}
        passthrough="attribute"
      >
        <div />
      </Component>
    </Fetcher(Unknown)>
  </Connect(Fetcher(Unknown))>
</Provider>
`;

exports[`Fetcher should mount, call fetch and dispatch actions 2`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
]
`;

exports[`Fetcher should mount, call fetch and dispatch actions 3`] = `
Array [
  Array [
    "/api/endpoint?",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Fetcher should render empty div if store gets messed with 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Fetcher(Unknown))
    passthrough="attribute"
  >
    <Fetcher(Unknown)
      addResource={[Function]}
      endpoint="/api/endpoint"
      fetchAdditionalSuccess={[Function]}
      fetchError={[Function]}
      fetchSuccess={[Function]}
      method="GET"
      modifyResource={[Function]}
      name="NAME"
      passthrough="attribute"
      removeResource={[Function]}
      requestHeader={
        Object {
          "Authorization": "Basic dummy_token",
          "Content-Type": "application/json",
        }
      }
      requestResource={[Function]}
    >
      <div />
    </Fetcher(Unknown)>
  </Connect(Fetcher(Unknown))>
</Provider>
`;

exports[`Fetcher should use loadInitialRows with different orderBys 1`] = `
Array [
  Array [
    "/api/zero?limit=500&offset=0",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
  Array [
    "/api/zero?limit=500&offset=0&sort=columnKey",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Fetcher should use loadMoreRows 1`] = `
Array [
  Array [
    "/api/zero?limit=500&offset=0",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
  Array [
    "/api/zero?limit=150&offset=100",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`Fetcher should use loadMoreRows 2`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": "pKey",
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "acceptResponse": undefined,
    "data": Object {
      "pKey": Array [
        1,
        2,
        3,
        4,
      ],
    },
    "name": "NAME",
    "type": "@@fetcher/FETCH_SUCCESS",
  },
  Object {
    "acceptResponse": undefined,
    "data": Object {
      "pKey": Array [
        1,
        2,
        3,
        4,
      ],
    },
    "name": "NAME",
    "type": "@@fetcher/FETCH_ADDITIONAL_SUCCESS",
  },
]
`;

exports[`Fetcher should use pagination 1`] = `
Array [
  Array [
    "/api/zero?limit=500&offset=0",
    Object {
      "body": undefined,
      "credentials": "include",
      "headers": Object {
        "Authorization": "Basic dummy_token",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;
